### PR TITLE
QNX message passing programs

### PR DESCRIPTION
**Pull request by KinhTruongUnikie**
The program can be run only with the combination of --help(optional),  --message \<server name\> and --file \<file name\>(in any order), any other combinations will result exit error.
The main implementation sits in ipc_methods.c.

### CODE REVIEW

**Comment by KinhTruongUnikie** [comment]
After the PR are done, I will resolve the conflict before merge

**Review by RomainChaumontet — CHANGES_REQUESTED** [review]
Looks like it doesn't work.
image

**Comment by KinhTruongUnikie** [comment]
Hi @RomainChaumontet, can you share the file you used to test the code?

File: C_hands_on_programming/ipc_methods.c [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    while ((coid = name_open(info->argument_string, 0)) == -1)
+    {
+        sleep(1);
+    }
+    printf("Found the server connection!\n");
+
+    //Start message reading
+	fseek(read, 0, SEEK_END);
+	size = ftell(read);
+	fseek(read, 0, SEEK_SET);
```
Comment (RomainChaumontet) [inline C_hands_on_programming/ipc_methods.c:217]: fseek should be avoid to find the size of a file. You can use instead ```struct stat()```.

**Review by baovu-unikie — APPROVED** [review]

File: C_hands_on_programming/ipc_methods.c [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+	SETIOV(&siov[1], buffer, header.data_size);
+
+	status = MsgSendvs(coid, siov, 2, NULL, 0);
+	if (status == -1) {
+		perror("MsgSendv");
+		exit(EXIT_FAILURE);
+	}
+	free(buffer);
+
+	if (MsgSendPulse(coid, -1, EOF_PULSE_CODE, 0) == -1) {
```
Comment (RomainChaumontet) [inline C_hands_on_programming/ipc_methods.c:246]: I think you can just skip sending a pulse -> when the client died, the server will automatically receive a pulse saying the client is disconnected.

File: C_hands_on_programming/ipc_methods.c [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+				printf("Server receive a pulse code of %d and value of %x\n",
+						rbuf.pulse.code, rbuf.pulse.value.sival_int);
+			}
+			//printf("Server receive a pulse code of %d and value of %x\n", rbuf.pulse.code, rbuf.pulse.value.sival_int);
+		} else {
+			if (rbuf.type == HEADER_IOV_TYPE) {
+				char *data = NULL;
+				FILE *write;
+				// open file for writing
+				write = fopen(info->filename, "wb");
```
Comment (RomainChaumontet) [inline C_hands_on_programming/ipc_methods.c:295]: I think you can improve the name of your variable. You use write to the fdes of the file to write in, but write is a verb and a POSIX function. I suggest something like file2write or something like that.

**Review by RomainChaumontet — APPROVED** [review]
Great work. Just minors comments.

File: C_hands_on_programming/ipc_methods.c (hunk 1 above)
Comment (KinhTruongUnikie) [reply C_hands_on_programming/ipc_methods.c:217]: Why is fseek not recommended?

File: C_hands_on_programming/ipc_methods.c (hunk 3 above)
Comment (KinhTruongUnikie) [reply C_hands_on_programming/ipc_methods.c:295]: Yeah I am aware of that, I just wanna include a MsgSendPulse from the client side for practice purpose

File: C_hands_on_programming/ipc_methods.c (hunk 2 above)
Comment (KinhTruongUnikie) [reply C_hands_on_programming/ipc_methods.c:246]: Yeah I am aware of that, I just wanna include a MsgSendPulse from the client side for practice purpose

File: C_hands_on_programming/ipc_methods.c (hunk 1 above)
Comment (RomainChaumontet) [reply C_hands_on_programming/ipc_methods.c:217]: https://wiki.sei.cmu.edu/confluence/display/c/FIO19-C.+Do+not+use+fseek()+and+ftell()+to+compute+the+size+of+a+regular+file 
Farnood have sent this link in the team chat/

**Pull request merged**